README.md
---
layout: home
title: "bezeye - Bézier curve texture+movement and Spatial Color Quantization"
image: assets/favimage-472x472.jpg
---

![teaser](https://xyzzy.github.io/bezeye-media/bezeye-840x472-9-5x5.gif)
\[24bit-colour converted to 9 colour palette\]

# bezeye

Bézier curve texture+movement and Spatial Color Quantization

Instead of an ad with tracking...  
Like to [donate ![opencollective](assets/opencollective-icon.svg)](https://opencollective.com/xyzzy1) some appreciation for the use or inspiration this gives you?

### Welcome to the Wonderful World of Bézier animations

*Capturing the essence of colour perception looking at the real thing*

Bezeye is an animated GIF showcasing Bézier Curve Texture+Movement and Spatial Color Quantization.  
It is also a contribution entry for the Evoke 2014 demoscene party [https://www.evoke.eu/2014](https://www.evoke.eu/2014).  
With size limitations, Bezeye tries to find the optimum between maximizing the number of frames and reducing the number of colours in a frame.  
End result are 60 frames at 20fps resulting in a 20M animated GIF.  
This project contains the modernised snapshot of the source for the code+tools, artwork and settings to create the GIF.

## Table of contents

  - [Welcome to the Wonderful World of Bezeye](#welcome-to-the-wonderful-world-of-bézier-animations)
  - [Table of contents](#table-of-contents)
  - [Showcasing](#showcasing)
    - [Bézier curves](#b%C3%A9zier-curves)
    - [Wings](#wings)
    - [Colour reduction](#colour-reduction)
    - [Storyboard](#storyboard)
  - [Original entry](#original-entry)
  - [Requirements](#requirements)
  - [Building](#building)
    - [Automatic building](#automatic-building)
    - [Manual building](#manual-building)
  - [Manifest](#manifest)
  - [Installation](#installation)
  - [Versioning](#versioning)
  - [License](#license)
  - [Acknowledgments](#acknowledgments)

##  Showcasing

The animation shows three floating eyes, each checking you out, dragging and confusing your attention with wide-spectrum colours.  
Movement is cyclic to make the replay a seamless continuation.  
The initial size of 1920x1080 was too demanding for the compo machine and was reduced to 1280x720.  
Using a tiny colour palette containing 7 colours that captures the perception of the variety of colours used.  
Outline and texture of the wings are Bézier curve based.  
Choreography is storyboard driven and stored as Bézier curves.

### Bézier curves

Natural movement is always in motions, direction changes gradually and never abrupt.  
Bézier curves are cubic polynomials that create patterns found in nature.  
The curve algorithms are from the Closed Continuous Bézier Curve project `ccbc` that re-masters the maths.  
Major enhancements are on-curve control points and ultra high-speed control point encoding.  
The `ccbc` project can be found here: [https://github.com/RockingShip/ccbc](https://github.com/RockingShip/ccbc)

Below is a small interactive demonstration of a 10-point re-mastered curve:  
Click on the image to jump to the containing page.  
The red pads can be dragged to change the curve.  
The green lines illustrate where the traditional off-curve control points are located.

\[click on image to show demonstration\]  
[![Closed Continuous Bézier Curves](assets/remastered-300x300.png)](https://rockingship.github.io/ccbc/remastered.html)

### Wings

Wing texture is created by rendering a Bézier curve and moving its on-curve control points.  
The control points move in a cyclic fashion allowing a seamless loop.  
Each rendering of the curve is done with a rainbow colouring that cycles.  
Frames are transparent, the animation needs to run an extra cycle before they become seamless.  
The animation is 60 frames long and 120 frames are created.

Below is an image of the base curve, click on the image to jump to the animation.

\[click on image to show animation\]  
[![Wings](assets/wing-237x300.png)](https://xyzzy.github.io/bezeye-media/wing.gif)

### Colour reduction

Colour reduction improves the compression used in GIF images.  
Spatial Color Quantization (`scq`) is used to reduce the palette to 7 colours.  
`scq` dithers an image using a smaller replacement palette.  
The colour average of the dithered image matches the average of the original.  
A `3x3` or `5x5` grid is used t calculate the average of the center pixel.       

Below is link to a comparison page displaying different palette sizes ranging from 4 to 16 colours.  
The animations are split-screen, the lower left being the original 24bit-colour and upper right being the colour reduced.

\[click on image to show demonstration\]  
[![Colour Reduction](assets/sbs-534x300-7.png)](Compare.html)

### Storyboard

The storyboard describes the eyes' choreography.  
It displays movement and 8 attributes as a seamless loop.  
Choreography tracks are closed continuous Bézier curves.  
Attention has been put into not making the wings collide when they pass each other.

Below is a link to the storyboard.  
At the bottom of the page are the Bézier control points that are manually cut/pasted into the rendering code. 

\[click on image to show demonstration\]  
[![Storyboard](assets/storyboard-534x300.png)](https://xyzzy.github.io/bezeye/storyboard.html)

## Original entry

The original entry was a collaboration with Hopscotch who supplied the Bézier outlines and storyboard settings.  
It differs with this project in that it uses a different 7-colour palette.  
Sadly I cannot reconstruct its creation.

Click on the image below to open the animation:

\[click on image to show animation\]  
[![Original](assets/bezeye-534x300-7-original.png)](https://xyzzy.github.io/bezeye-media/bezeye-1280x720-7-original.gif)

## Requirements

  * `gcc` or compatible
  * `libgd` for 2d graphics support [https://libgd.github.io](https://libgd.github.io)
  * `mootools` for the scoreboard GUI [https://mootools.net](https://mootools.net)
  * `php` for scripts
  * `ffmpeg` (optional) for comparison

## Building

```sh
        ./autogen.sh  
        ./configure  
        make
```

### Automatic building

Make has a number of prime targets:

```
        bezeye-840x472-<numCol>-3x3.gif  
        bezeye-840x472-<numCol>-5x5.gif  
        bezeye-1280x720-<numCol>-3x3.gif  
        bezeye-1280x720-<numCol>-5x5.gif  
        bezeye-1280x720-7-original.gif  
        all-gif    (all gifs)  
        all-sbs    (all side-by-sides)  
        media      (all media files for this project)  
        
        3x3/5x5 are to indicate the grid size for determining the colour average.
```

### Manual building

Below are conceptual instructions for generating the animation.  
The file [Build.txt](Build.txt) contains detailed instructions for different sizes and palettes.

  - Load the storyboard into a browser and tweak accordingly.  
    Copy the `JSON` data to `render.php`.

  - Create wing texture  
    `mkdir wingdata`  
    `php wing.php wingdata/%02d.png`  
    This will generate 3 cycles of 60 frames that should animate as seamless.

  - Create 120 24bit-colour frames for animation (hardcoded 1280x720)  
    `mkdir frames-1280x720`  
    `php render.php wingdata/%02d.png frames-1280x720/%03d.png`

  - Extract 7 colour palette from first frame  
    `./scq6 --palette=octree --filter=5 --genpalette frames-1280x720/000.png - <numColour> /dev/null >palette-<numCol>.pal`

  - Colour compress frames  
    ./gen-bezeye.sh

  - collect parts for animated gifs  
    ./mergeraw bezeye-1280x720-7.gif bezeye-1280x720-7/img-000.gif bezeye-1280x720-7/diff-???.gif --delay=5 -f

## Manifest

  - [Build.txt](Build.txt)  
    Manual build instructions

  - [diffgif.cc](diffgif.cc)  
    GIF transparency encoder for frames

  - [gen-bezeye.sh](gen-bezeye.sh)  
    Script for colour compressing frames

  - [mergeraw.cc](mergeraw.cc)  
    Create animated GIF from individual frames

  - [mootools-*.js](mootools-core-1.4.5.js)  
    Mootools framework for [storyboard.html](storyboard.html)

  - [original-7.pal](original-7.pal)  
    Original entry 7-colour palette.

  - [render.php](render.php)  
    Render 24bit-colour frames

  - [scq6.cc](scq6.cc)  
    Spatial Color Quantization, convert 24bit-colour to palette

  - [storyboard.html](storyboard.html)  
    Storyboard GUI. Paste the settings in `render.php`

  - [wing.php](wing.php)  
    Render 24bit-colour bézier based texture for wings

## Installation

The file [tools/Build.txt](Build.txt) contains manual instructions to re-create the competition entry.

This project focuses on a single use-case with no consideration towards portability.

## Source code

Grab one of the tarballs at [https://github.com/xyzzy/bezeye/releases](https://github.com/xyzzy/bezeye/releases) or checkout the latest code:

```sh
  git clone https://github.com/xyzzy/bezeye.git
```

## Versioning

Using [SemVer](http://semver.org/) for versioning. For the versions available, see the [tags on this repository](https://github.com/xyzzy/bezeye/tags).

## License

This project is licensed under the GNU General Public License v3 - see the [LICENSE.txt](LICENSE.txt) file for details

## Acknowledgments

* Hopscotch for the amazing analog input and storyboard settings
* Derrick Coetzee and his amazing work on Spatial Color Quantization.
* [https://github.com/RockingShip/ccbc](https://github.com/RockingShip/ccbc) project for the amazing work on Closed Continuous Bézier Curves
* [https://github.com/xyzzy/jsGifEncoder](https://github.com/xyzzy/jsGifEncoder) project for the high-speed inline GIF encoding
* [https://github.com/xyzzy/scq](https://github.com/xyzzy/scq) project for the enhanced Spatial Color Quantization implementation


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

```
2021-03-26 18:18:28 Added: Donate button.
2021-03-23 17:20:40 Changed: Order items in navbar.
2021-03-21 16:13:11 Added: Reference to `bezeye-media` in navbar.
```

## 2021-03-21 00:07:22 [Version 1.1.0]

```
2021-03-21 14:56:06 Fixed: Added `assets/remastered-300x300.png`.
2021-03-19 17:43:15 Added: `moonwalk` theme.
2021-02-17 19:26:24 Changed: Table of contents and Headings in `README.md`.
2021-01-22 14:17:51 Changed: Italic sub-title.
```

## 2021-01-21 02:52:54 Version 1.0.0

```
2021-01-21 02:48:11 Changed: "true-colour" -> "24bit-colour".
2021-01-21 00:48:48 Added: `bezeye-840x472-9-5x5.gif`.
2021-01-21 00:07:52 Changed: Updated links to `bezeye-media`.
2021-01-20 18:02:12 Changed: Makefile and tools more file/path aware.
2021-01-20 17:55:34 Added: Recovered lost octree code.
2021-01-12 11:56:13 Historic commit.
```

[Unreleased]: https://github.com/xyzzy/bezeye/compare/v1.1.0...HEAD
[Version 1.1.0]: https://github.com/xyzzy/bezeye/compare/v1.0.0...v1.1.0


wing.php
<?php

/*
 * This file is part of bezeye, Evoke 2014 demoscene party entry
 * Copyright (C) 2014, xyzzy@rockingship.net
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

if (empty($argv[1]) || empty($argv[2]))
    die ('usage: '.$argv[0]. ' <wingOutline> <fileTemplate>'."\n");

$v0 = 0.65; // 0.55
$a0 = 0.01;
ini_set('memory_limit','3g');

function rgb($ang, $amp=1)
{
global $im;

	$ang = $ang-floor($ang);
	$ang *= 6;

		if ($ang < 1) {
			$r = 1;
			$g = $ang-0;
			$b = 0;
		} else if ($ang < 2) {
			$r = 2-$ang;
			$g = 1;
			$b = 0;
		} else if ($ang < 3) {
			$r = 0;
			$g = 1;
			$b = $ang-2;
		} else if ($ang < 4) {
			$r = 0;
			$g = 4-$ang;
			$b = 1;
		} else if ($ang < 5) {
			$r = $ang-4;
			$g = 0;
			$b = 1;
		} else {
			$r = 1;
			$g = 0;
			$b = 6-$ang;
		}

	$r *= 255;
$r = 255-(255-$r)*$amp;
	$g *= 255;
$g = 255-(255-$g)*$amp;
	$b *= 255;
$b = 255-(255-$b)*$amp;
	return imagecolorallocate($im, $r,$g,$b);
}

$_x = array();
$_y = array();
$_dx = array();
$_dy = array();
$_a = array();
$_v = array();
$_c = array();

$im0 = imagecreatefrompng($argv[1]);
if (!$im0)
    die ("wing outline not found\n");

$im = imagecreatetruecolor(792,1000);
imagesavealpha($im, true);
imagealphablending($im, false);
$white = imagecolorallocate($im, 255,255,255);
$black = imagecolorallocate($im, 0,0,0);
$green = imagecolorallocate($im, 0,255,0);
$blue = imagecolorallocate($im, 0,0,255);
$transp = imagecolorallocatealpha($im, 255,255,255,127);

//--- beginInit
$maxslice = 100;
$wmin = 50;
$wmax = 130;
$m = 10;
$coef = array(56.0/209.0, -15.0/209.0, 4.0/209.0, -1.0/209.0, 0, 1.0/209.0, -4.0/209.0, 15.0/209.0, -56.0/209.0, 1 );
$Ax = array (382,423,414,365,236,131,214, 88,100,240);
$Ay = array ( 80,209,362,421,464,347,263,178, 67, 16);

if (1) {
// 30
$Ax = array(411,421,423,423,422,416,401,365,328,286,242,200,165,138,133,162,200,203,164,125,88,67,78,106,144,186,231,275,318,357);
$Ay = array(128,172,217,262,307,352,394,420,446,463,465,449,421,385,342,308,284,248,225,204,178,139,96,61,36,21,16,21,36,58);
$summax = 1377.088348;
} else if (1) {
//60
$Ax = array(402,411,417,421,422,423,424,423,422,421,418,415,409,397,379,360,342,322,301,279,257,234,213,193,175,158,145,134,130,138,153,172,191,209,210,192,172,151,132,112,93,78,68,68,75,87,102,120,140,161,182,205,228,250,273,295,316,337,356,374);
$Ay = array(108,129,151,173,196,219,242,265,288,311,333,356,378,397,411,424,437,449,459,465,467,463,455,444,430,414,396,376,353,332,315,302,290,276,255,241,230,219,208,196,183,166,145,123,101,82,65,50,38,29,22,18,16,17,20,26,35,45,57,72);
$summax = 1381.443780;
} else {
// 120
$Ax = array(389,396,402,407,411,414,417,419,421,422,422,423,423,424,424,423,423,423,422,422,421,420,418,417,415,412,408,403,395,386,377,367,358,348,339,329,319,308,298,287,275,264,252,241,230,219,208,198,189,180,171,163,155,148,142,136,132,130,131,135,141,149,158,168,177,187,197,206,213,212,205,195,186,176,165,155,145,135,125,115,105,96,87,80,74,69,67,67,70,74,79,85,93,101,109,118,128,138,148,159,170,181,192,203,215,226,238,249,261,272,283,294,305,316,326,337,346,356,365,374);
$Ay = array(89,98,108,119,129,140,152,163,174,186,197,209,220,232,243,255,267,278,290,301,313,324,336,347,358,370,380,391,399,406,413,419,426,433,439,445,451,456,460,463,466,467,467,465,462,458,453,448,441,434,427,419,410,401,391,381,370,359,347,337,327,319,311,305,298,292,286,279,270,259,250,243,237,232,226,221,216,210,204,198,192,185,177,169,159,148,137,126,114,103,93,84,75,66,59,52,45,39,34,30,26,22,20,18,16,16,16,17,18,20,23,26,30,35,40,45,51,57,64,72);
$summax = 1383.186163;
}
$m = count($Ax);
$gen = 0;
printf("count(Ax):%d\n", count($Ax));

// rescale
if (1) {
	$xmin =  66; // 66.709622887307-0.0001;
	$xmax = 424; // 423.24696551749+0.0001;
	$ymin =  15; // 15.978061786642-0.0001;
	$ymax = 467; // 466.30044180059+0.0001;
	$scale = 1000/($ymax-$ymin);
$scale = 1;
	for ($i=0; $i<$m; $i++) {
		$Ax[$i] = ($Ax[$i]-$xmin)*$scale;
		$Ay[$i] = ($Ay[$i]-$ymin)*$scale;
		$Ax[$i] = round($Ax[$i]);
		$Ay[$i] = round($Ay[$i]);
	}
	echo 'Ax[] = '.json_encode($Ax)."\n";
	echo 'Ay[] = '.json_encode($Ay)."\n";

$scale = 1000/($ymax-$ymin);
}

// determine control points
$Bx = array();
$By = array();
for ($i=0; $i<$m; $i++) {

	$Bx[$i] = $Ax[$i]* 1.000000000000
		+$Ax[($i+1)%$m]* 0.267942583732+$Ax[($i+2)%$m]*-0.071770334928+$Ax[($i+3)%$m]* 0.019138755981+$Ax[($i+4)%$m]*-0.004784688995
		+$Ax[($i-4+$m)%$m]* 0.004784688995+$Ax[($i-3+$m)%$m]*-0.019138755981+$Ax[($i-2+$m)%$m]* 0.071770334928+$Ax[($i-1+$m)%$m]*-0.267942583732;
	$By[$i] = $Ay[$i]* 1.000000000000
		+$Ay[($i+1)%$m]* 0.267942583732+$Ay[($i+2)%$m]*-0.071770334928+$Ay[($i+3)%$m]* 0.019138755981+$Ay[($i+4)%$m]*-0.004784688995
		+$Ay[($i-4+$m)%$m]* 0.004784688995+$Ay[($i-3+$m)%$m]*-0.019138755981+$Ay[($i-2+$m)%$m]* 0.071770334928+$Ay[($i-1+$m)%$m]*-0.267942583732;
}
$Cx = array();
$Cy = array();
for ($i=0; $i<$m; $i++) {
	$Cx[$i] = 2*$Ax[($i+1)%$m] - $Bx[($i+1)%$m];
	$Cy[$i] = 2*$Ay[($i+1)%$m] - $By[($i+1)%$m];
}

// walk the path to determine boundingbox (and $summax)
$xmin = $xmax = $Ax[0];
$ymin = $ymax = $Ay[0];
$d=1000; $xofs=0; $yofs=0;
$summax = 0;
for ($k=0; $k<$m; $k++) {

	$x0 = $Ax[$k];
	$y0 = $Ay[$k];
	$x1 = $Bx[$k];
	$y1 = $By[$k];
	$x2 = $Cx[$k];
	$y2 = $Cy[$k];
	$x3 = $Ax[($k+1)%$m];
	$y3 = $Ay[($k+1)%$m];

	$n = $maxslice;
	$lastx = $x0;
	$lasty = $y0;
	for ($i = 1; $i < $n; $i++) {
		$t = $i / $n;
		$t1 = 1 - $t;

		$ex = $x0+($x1-$x0)*$t;
		$ey = $y0+($y1-$y0)*$t;
		$fx = $x1+($x2-$x1)*$t;
		$fy = $y1+($y2-$y1)*$t;
		$gx = $x2+($x3-$x2)*$t;
		$gy = $y2+($y3-$y2)*$t;

		$hx = $ex+($fx-$ex)*$t;
		$hy = $ey+($fy-$ey)*$t;
		$ix = $fx+($gx-$fx)*$t;
		$iy = $fy+($gy-$fy)*$t;

		$x = $hx+($ix-$hx)*$t;
		$y = $hy+($iy-$hy)*$t;

		if ($x < $xmin) $xmin = $x; else if ($x > $xmax) $xmax = $x;
		if ($y < $ymin) $ymin = $y; else if ($y > $ymax) $ymax = $y;

//		if (abs($y-500) < $d && $x > $xofs) {
//			$d = abs($y-500);
//			$xofs = $x;
//		}
		if ($x > $xofs) {
			$xofs = $x;
			$yofs = $y;
		}

		$lastx -= $x;
		$lasty -= $y;
		$summax += sqrt($lastx*$lastx+$lasty*$lasty);
		$lastx = $x;
		$lasty = $y;

	}	
}
echo "$xmin $xmax $ymin $ymax | $xofs $yofs\n";

//--- endInit

for ($outer=0; $outer<3; $outer++) 
for ($loop=0; $loop<60; $loop++) {

echo "$outer.$loop\n";

// storage for gradient lines
$gx0 = array();
$gy0 = array();
$gx1 = array();
$gy1 = array();

	foreach ($_x as $i => $v) {

		imagearc($im, $_x[$i]*$scale,$_y[$i]*$scale, 4,4, 0,360, rgb($_c[$i], ($_v[$i]*$_v[$i])/($v0*$v0)));
		$_x[$i] += $_dx[$i]*$_v[$i];
		$_y[$i] += $_dy[$i]*$_v[$i];
		$_v[$i] -= $a0;
		$_a[$i] -= 0.1;
		if ($_v[$i] < 0) $_v[$i] = 0;
		if ($_a[$i] < 0) $_a[$i] = 0;
	}

// plot bezier

$sum = 0;

for ($k=0; $k<$m; $k++) {

	$x0 = $Ax[$k];
	$y0 = $Ay[$k];
	$x1 = $Bx[$k];
	$y1 = $By[$k];
	$x2 = $Cx[$k];
	$y2 = $Cy[$k];
	$x3 = $Ax[($k+1)%$m];
	$y3 = $Ay[($k+1)%$m];

	$n = $maxslice;

	$lastx = $x0;
	$lasty = $y0;
	for ($i = 1; $i < $n; $i++) {
		$t = $i / $n;
		$t1 = 1 - $t;

/*
		$a = $t1*$t1;
		$d = $t*$t;
		$b = $t*$a*3;
		$c = $t1*$d*3;
		$a *= $t1;
		$d *= $t;

		$x = $a*$x0 + $b*$x1 + $c*$x2 + $d*$x3;
		$y = $a*$y0 + $b*$y1 + $c*$y2 + $d*$y3;
		$lastx -= $x;
		$lasty -= $y;

		$sum += sqrt($lastx*$lastx+$lasty*$lasty);

		$lastx = $x;
		$lasty = $y;
*/

		$ex = $x0+($x1-$x0)*$t;
		$ey = $y0+($y1-$y0)*$t;
		$fx = $x1+($x2-$x1)*$t;
		$fy = $y1+($y2-$y1)*$t;
		$gx = $x2+($x3-$x2)*$t;
		$gy = $y2+($y3-$y2)*$t;

		$hx = $ex+($fx-$ex)*$t;
		$hy = $ey+($fy-$ey)*$t;
		$ix = $fx+($gx-$fx)*$t;
		$iy = $fy+($gy-$fy)*$t;

		$x = $hx+($ix-$hx)*$t;
		$y = $hy+($iy-$hy)*$t;

		imagesetpixel($im, $x*$scale, $y*$scale, $black);
		$gx0[] = $x;
		$gy0[] = $y;

		$lastx -= $x;
		$lasty -= $y;
		$sum += sqrt($lastx*$lastx+$lasty*$lasty);
		$lastx = $x;
		$lasty = $y;


		$dx = $ix-$hx;
		$dy = $iy-$hy;

		$w = (1+sin(2*M_PI*$loop/60+6*2*M_PI*$sum/$summax))*.5*($wmax-$wmin)+$wmin;

		$xx = $x - $dy*$w/sqrt($dx*$dx+$dy*$dy);
		$yy = $y + $dx*$w/sqrt($dx*$dx+$dy*$dy);

if(1)
		imagesetpixel($im, $xx*$scale, $yy*$scale, $green);
		$gx1[] = $xx;
		$gy1[] = $yy;

if (1) {
		$_x[] = $xx;
		$_y[] = $yy;
		$_dx[] = $dx;
		$_dy[] = $dy;
		$_a[] = 0.9;
		$_v[] = $v0;
		$_c[] = $sum/$summax+$loop/60.0;
}
	}

}
printf("summax:%f\n", $sum);

if(1)
	for ($y=0; $y<1000; $y++) {
		for ($x=0; $x<792; $x++) {
			$c = imagecolorat($im0,$x,$y);
			if (($c>>24) != 0)
				imagesetpixel($im, $x, $y, $transp);
		}
	}

imagepng($im, sprintf($argv[2], $loop));

}
